Trying to use a long-lived certificate with the wrong/incorrect system login makes tsh segfault

**What happened**:
Issuing a certificate with `tctl auth sign` on an auth server and then trying to use that certificate for `tsh ssh` with an incorrect login username makes `tsh` segfault.

**What you expected to happen**:
The error message from `tsh` should say `access denied to test@server` or similar, rather than segfaulting.

**How to reproduce it (as minimally and precisely as possible)**:
For information: the `logins` for this user are actually not specified, but the user has the `clusteradmin` role which grants login as the `root` user - not sure if this is relevant here but thought I'd include it in case it's important.

User:
```
root@gus-main-auth-0:/# tctl get user/test
kind: user
metadata:
  id: 1569865631292691176
  name: test
spec:
  created_by:
    time: 0001-01-01T00:00:00Z
    user:
      name: ""
  expires: 0001-01-01T00:00:00Z
  roles:
  - clusteradmin
  status:
    is_locked: false
    lock_expires: 0001-01-01T00:00:00Z
    locked_time: 0001-01-01T00:00:00Z
  traits:
    kubernetes_groups: null
    logins: null
version: v2
```

Role:
```
root@gus-main-auth-0:/# tctl get role/clusteradmin
kind: role
metadata:
  id: 1569863314494056944
  name: clusteradmin
spec:
  allow:
    kubernetes_groups:
    - system:masters
    logins:
    - root
    node_labels:
      '*': '*'
    rules:
    - resources:
      - '*'
      verbs:
      - '*'
  deny:
    logins: null
  options:
    cert_format: standard
    forward_agent: true
    max_session_ttl: 12h0m0s
    port_forwarding: true
version: v3
```

To repro - issue a long-lived certificate using `tctl` on the auth server:
```
root@gus-main-auth-0:/# tctl auth sign --ttl=14800h --user=test --out=test.pem

The certificate has been written to test.pem
```

Copy that certificate to the local machine, then use it with a local `tsh` client (with a system username that is not `root`) and observe the segfault:
```
$ tsh --debug ssh -i test.pem --proxy=gus-main.gravitational.co:3080 test@gus-main-auth-0                                                                                                                                                                                                                gus@antaeus
INFO [TSH]       Reading identity file: test.pem tsh/tsh.go:1002
DEBU [TSH]       Extracted username "test" from the identity file test.pem. tsh/tsh.go:864
INFO [CLIENT]    Connecting proxy=gus-main.gravitational.co:3023 login='test' method=0 client/api.go:1465
DEBU [CLIENT]    Activating relogin on ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain. client/api.go:306
DEBU [CLIENT]    not using loopback pool for remote proxy addr: gus-main.gravitational.co:3080 client/api.go:1883
DEBU [CLIENT]    HTTPS client init(proxyAddr=gus-main.gravitational.co:3080, insecure=false) client/weblogin.go:252
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0xc6337c]

goroutine 1 [running]:
github.com/gravitational/teleport/lib/client.(*LocalKeyAgent).UpdateProxyHost(...)
	/gopath/src/github.com/gravitational/teleport/lib/client/keyagent.go:118
github.com/gravitational/teleport/lib/client.(*TeleportClient).applyProxySettings(0xc00000d0e0, 0x1, 0xc0005084a0, 0x1e, 0xc00059c6f0, 0xc, 0xc00059c700, 0xc, 0xc0005084e0, 0x1e, ...)
	/gopath/src/github.com/gravitational/teleport/lib/client/api.go:1702 +0x8cc
github.com/gravitational/teleport/lib/client.(*TeleportClient).Login(0xc00000d0e0, 0x1843a00, 0xc0000df980, 0xc0002f1901, 0x1, 0x1, 0x0)
	/gopath/src/github.com/gravitational/teleport/lib/client/api.go:1530 +0x191
github.com/gravitational/teleport/lib/client.RetryWithRelogin(0x1843a00, 0xc0000df980, 0xc00000d0e0, 0xc0002f1a00, 0x0, 0x1)
	/gopath/src/github.com/gravitational/teleport/lib/client/api.go:306 +0x12e
main.onSSH(0xc000589200)
	/gopath/src/github.com/gravitational/teleport/tool/tsh/tsh.go:710 +0xd7
main.Run(0xc0000e0010, 0x6, 0x6, 0x0)
	/gopath/src/github.com/gravitational/teleport/tool/tsh/tsh.go:325 +0x5183
main.main()
	/gopath/src/github.com/gravitational/teleport/tool/tsh/tsh.go:170 +0x11c
```

Run the same command again (but use the correct, allowed `root` user this time) and see that the command works as intended:
```
$ tsh --debug ssh -i test.pem --proxy=gus-main.gravitational.co:3080 root@gus-main-auth-0
INFO [TSH]       Reading identity file: test.pem tsh/tsh.go:1002
DEBU [TSH]       Extracted username "test" from the identity file test.pem. tsh/tsh.go:864
INFO [CLIENT]    Connecting proxy=gus-main.gravitational.co:3023 login='root' method=0 client/api.go:1465
INFO [CLIENT]    Successful auth with proxy gus-main.gravitational.co:3023 client/api.go:1471
DEBU [CLIENT]    Found clusters: [{"name":"gus-main.gravitational.co","lastconnected":"2019-09-30T17:58:25.347073669Z","status":"online"},{"name":"gus-east.gravitational.co","lastconnected":"2019-09-30T17:53:36.13399115Z","status":"online"}] client/client.go:104
INFO [CLIENT]    Client= connecting to node=gus-main-auth-0:0@default@gus-main.gravitational.co client/client.go:363
root@gus-main-auth-0:~#
```

**Environment**:
- Teleport version (use `teleport version`): `Teleport Enterprise v4.0.4git:v4.0.4-0-g1a2ed507 go1.12.1
`
- Tsh version (use `tsh version`): `Teleport v4.0.2 git:v4.0.2-0-gb7e0e872 go1.12.1`
- OS (e.g. from /etc/os-release): `Fedora 29`

--- a/lib/client/api.go
+++ b/lib/client/api.go
@@ -1509,8 +1509,8 @@ func (tc *TeleportClient) getProxySSHPrincipal() string {
 		proxyPrincipal = tc.JumpHosts[0].Username
 	}
 	// see if we already have a signed key in the cache, we'll use that instead
-	if !tc.Config.SkipLocalAuth && tc.LocalAgent() != nil {
-		signers, err := tc.LocalAgent().Signers()
+	if !tc.Config.SkipLocalAuth && tc.localAgent != nil {
+		signers, err := tc.localAgent.Signers()
 		if err != nil || len(signers) == 0 {
 			return proxyPrincipal
 		}
@@ -1526,8 +1526,8 @@ func (tc *TeleportClient) getProxySSHPrincipal() string {
 // can use to try to authenticate
 func (tc *TeleportClient) authMethods() []ssh.AuthMethod {
 	m := append([]ssh.AuthMethod(nil), tc.Config.AuthMethods...)
-	if tc.LocalAgent() != nil {
-		m = append(m, tc.LocalAgent().AuthMethods()...)
+	if tc.localAgent != nil {
+		m = append(m, tc.localAgent.AuthMethods()...)
 	}
 	return m
 }
@@ -1614,8 +1614,10 @@ func (tc *TeleportClient) connectToProxy(ctx context.Context) (*ProxyClient, err
 // Logout removes certificate and key for the currently logged in user from
 // the filesystem and agent.
 func (tc *TeleportClient) Logout() error {
-	err := tc.localAgent.DeleteKey()
-	if err != nil {
+	if tc.localAgent == nil {
+		return nil
+	}
+	if err := tc.localAgent.DeleteKey(); err != nil {
 		return trace.Wrap(err)
 	}
 
@@ -1625,8 +1627,10 @@ func (tc *TeleportClient) Logout() error {
 // LogoutAll removes all certificates for all users from the filesystem
 // and agent.
 func (tc *TeleportClient) LogoutAll() error {
-	err := tc.localAgent.DeleteKeys()
-	if err != nil {
+	if tc.localAgent == nil {
+		return nil
+	}
+	if err := tc.localAgent.DeleteKeys(); err != nil {
 		return trace.Wrap(err)
 	}
 	return nil
@@ -1728,7 +1732,7 @@ func (tc *TeleportClient) Login(ctx context.Context, activateKey bool) (*Key, er
 	key.ClusterName = response.HostSigners[0].ClusterName
 	tc.SiteName = response.HostSigners[0].ClusterName
 
-	if activateKey {
+	if activateKey && tc.localAgent != nil {
 		// save the list of CAs client trusts to ~/.tsh/known_hosts
 		err = tc.localAgent.AddHostSignersToCache(response.HostSigners)
 		if err != nil {
@@ -1782,6 +1786,9 @@ func (tc *TeleportClient) GetTrustedCA(ctx context.Context, clusterName string) 
 // UpdateTrustedCA connects to the Auth Server and fetches all host certificates
 // and updates ~/.tsh/keys/proxy/certs.pem and ~/.tsh/known_hosts.
 func (tc *TeleportClient) UpdateTrustedCA(ctx context.Context, clusterName string) error {
+	if tc.localAgent == nil {
+		return trace.BadParameter("TeleportClient.UpdateTrustedCA called on a client without localAgent")
+	}
 	// Get the list of host certificates that this cluster knows about.
 	hostCerts, err := tc.GetTrustedCA(ctx, clusterName)
 	if err != nil {
@@ -1833,10 +1840,12 @@ func (tc *TeleportClient) applyProxySettings(proxySettings ProxySettings) error 
 		}
 		tc.WebProxyAddr = net.JoinHostPort(addr.Host(), strconv.Itoa(addr.Port(defaults.HTTPListenPort)))
 
-		// Update local agent (that reads/writes to ~/.tsh) with the new address
-		// of the web proxy. This will control where the keys are stored on disk
-		// after login.
-		tc.localAgent.UpdateProxyHost(addr.Host())
+		if tc.localAgent != nil {
+			// Update local agent (that reads/writes to ~/.tsh) with the new address
+			// of the web proxy. This will control where the keys are stored on disk
+			// after login.
+			tc.localAgent.UpdateProxyHost(addr.Host())
+		}
 	}
 	// Read in settings for the SSH endpoint of the proxy.
 	//
@@ -1891,7 +1900,10 @@ func (tc *TeleportClient) localLogin(ctx context.Context, secondFactor string, p
 
 // AddTrustedCA adds a new CA as trusted CA for this client, used in tests
 func (tc *TeleportClient) AddTrustedCA(ca services.CertAuthority) error {
-	err := tc.LocalAgent().AddHostSignersToCache(auth.AuthoritiesToTrustedCerts([]services.CertAuthority{ca}))
+	if tc.localAgent == nil {
+		return trace.BadParameter("TeleportClient.AddTrustedCA called on a client without localAgent")
+	}
+	err := tc.localAgent.AddHostSignersToCache(auth.AuthoritiesToTrustedCerts([]services.CertAuthority{ca}))
 	if err != nil {
 		return trace.Wrap(err)
 	}
@@ -1899,7 +1911,7 @@ func (tc *TeleportClient) AddTrustedCA(ca services.CertAuthority) error {
 	// only host CA has TLS certificates, user CA will overwrite trusted certs
 	// to empty file if called
 	if ca.GetType() == services.HostCA {
-		err = tc.LocalAgent().SaveCerts(auth.AuthoritiesToTrustedCerts([]services.CertAuthority{ca}))
+		err = tc.localAgent.SaveCerts(auth.AuthoritiesToTrustedCerts([]services.CertAuthority{ca}))
 		if err != nil {
 			return trace.Wrap(err)
 		}
@@ -1908,8 +1920,11 @@ func (tc *TeleportClient) AddTrustedCA(ca services.CertAuthority) error {
 	return nil
 }
 
-// AddKey adds a key to the client's local agent
+// AddKey adds a key to the client's local agent, used in tests.
 func (tc *TeleportClient) AddKey(host string, key *Key) (*agent.AddedKey, error) {
+	if tc.localAgent == nil {
+		return nil, trace.BadParameter("TeleportClient.AddKey called on a client without localAgent")
+	}
 	return tc.localAgent.AddKey(key)
 }
 

